### PR TITLE
New module: na_ontap_net_subnet

### PR DESCRIPTION
**Pull request by albinpopote**
##### SUMMARY

New NetApp module to manage network subnets (create, modify, rename and delete actions)

##### ISSUE TYPE

- New Module Pull Request

##### COMPONENT NAME

na_ontap_net_subnet

##### ADDITIONAL INFORMATION

Tested on FAS8200 with DOT version 9.4 : 

Playbook used:

```paste below
---

- hosts:
  - OPVSIMPP01
  tasks:
    - name: Subnet Create
      na_ontap_net_subnet:
        name: "subnet-adm"
        broadcast_domain: "Default"
        subnet: "172.17.133.0/24"
        ip_ranges: [ '172.17.133.210-172.17.133.215', '172.17.133.225' ]
        gateway: "172.17.133.1"
        ipspace: "Default"
        state: present
        hostname: "{{ hostname }}"
        username: "{{ username }}"
        password: "{{ password }}"
        http_port: "{{ http_port | default( 443) }}"
        https: "{{ https | default( true) }}"
        validate_certs: "{{ validate_certs | default( false) }}"

    - name: Subnet Rename
      na_ontap_net_subnet:
        name: "subnet-adm-new"
        from_name: "subnet-adm"
        state: present
        hostname: "{{ hostname }}"
        username: "{{ username }}"
        password: "{{ password }}"
        http_port: "{{ http_port | default( 443) }}"
        https: "{{ https | default( true) }}"
        validate_certs: "{{ validate_certs | default( false) }}"

    - name: Subnet Modify
      na_ontap_net_subnet:
        name: "subnet-adm-new"
        ip_ranges: [ '172.17.133.210-172.17.133.216', '172.17.133.225' ]
        state: present
        hostname: "{{ hostname }}"
        username: "{{ username }}"
        password: "{{ password }}"
        http_port: "{{ http_port | default( 443) }}"
        https: "{{ https | default( true) }}"
        validate_certs: "{{ validate_certs | default( false) }}"

    - name: Subnet Delete
      na_ontap_net_subnet:
        state: absent
        name: "subnet-adm-new"
        hostname: "{{ hostname }}"
        username: "{{ username }}"
        password: "{{ password }}"
        http_port: "{{ http_port | default( 443) }}"
        https: "{{ https | default( true) }}"
        validate_certs: "{{ validate_certs | default( false) }}"

  gather_facts: no
```

I will try to code a test-unit this week-end or the next week.

### CODE REVIEW

File: lib/ansible/modules/storage/netapp/na_ontap_net_subnet.py [hunk 1]
```diff
@@ -0,0 +1,324 @@
...
+
+
+ANSIBLE_METADATA = {'metadata_version': '1.1',
+                    'status': ['preview'],
+                    'supported_by': 'certified'}
+
+DOCUMENTATION = """
+module: na_ontap_net_subnet
+short_description: NetApp Ontap Create, delete, modify network subnets.
+extends_documentation_fragment:
```
Comment (lonico) [inline lib/ansible/modules/storage/netapp/na_ontap_net_subnet.py:16]: nitpicks: ONTAP create

File: lib/ansible/modules/storage/netapp/na_ontap_net_subnet.py [hunk 2]
```diff
@@ -0,0 +1,324 @@
...
+  state:
+    description:
+    - Whether the specified network interface group should exist or not.
+    choices: ['present', 'absent']
+    default: present
+
+  broadcast_domain:
+    description:
+    - Specify the required broadcast_domain name for the subnet.
+    - A broadcast domain can not be modified after the subnet has been created
```
Comment (lonico) [inline lib/ansible/modules/storage/netapp/na_ontap_net_subnet.py:32]: nitpick: cannot

File: lib/ansible/modules/storage/netapp/na_ontap_net_subnet.py [hunk 3]
```diff
@@ -0,0 +1,324 @@
...
+HAS_NETAPP_LIB = netapp_utils.has_netapp_lib()
+
+
+class NetAppOntapSubnet(object):
+    """
+    Create, Modifies and Destroys a subnet
+    """
+    def __init__(self):
+        """
+        Initialize the Ontap Subnet class
```
Comment (lonico) [inline lib/ansible/modules/storage/netapp/na_ontap_net_subnet.py]: ONTAP

**Review by lonico — APPROVED** [review]

**Comment by angystardust** [comment]
thanks @albinpopote very useful :+1: 
@lonico @carchi8py Any reason for not merging this module? :thinking:

**Comment by carchi8py** [comment]
@albinpopote let me talk to RedHat, they control merging for Ansible.

File: lib/ansible/modules/storage/netapp/na_ontap_net_subnet.py [hunk 4]
```diff
@@ -0,0 +1,324 @@
...
+EXAMPLES = """
+    - name: create subnet
+      na_ontap_net_subnet:
+        state: present
+        username: "{{ netapp_username }}"
+        password: "{{ netapp_password }}"
+        hostname: "{{ netapp_hostname }}"
+        subnet: 10.10.10.0/24
+        name: subnet-adm
+        ip_range: [ '10.10.10.30-10.10.10.40', '10.10.10.51' ]
```
Comment (angystardust) [inline lib/ansible/modules/storage/netapp/na_ontap_net_subnet.py]: little typo: the parameter name should be `ip_ranges`

**Comment by angystardust** [comment]
If it's useful for speed up the merging process, I've just tested this module on a NetApp Release 9.4P2 and works great! :+1: 
There's just a little typo in the example section...I've made a inline comment ;)

**Comment by albinpopote** [comment]
Thanks @angystardust for this code review. It's corrected.

@carchi8py I discussed with dblackwe on slack about merging and validation by RedHat. I know that the process is not very simple... I have others module in the pipe (and module bug fix too), I would like to know if i make a pull reqquest for them (and maybe not certified and not supported by the NetApp Team) or if I send you new modules to test and publish as NetApp Team to get supported and certified ?

Available on slack if you want to discuss on this.

**Review by lonico — APPROVED** [review]
shipit

File: lib/ansible/modules/storage/netapp/na_ontap_net_subnet.py [hunk 5]
```diff
@@ -0,0 +1,324 @@
+#!/usr/bin/python
+
+# (c) 2018, NetApp, Inc
```
Comment (gundalow) [inline lib/ansible/modules/storage/netapp/na_ontap_net_subnet.py]: Is this right?

**Comment by angystardust** [comment]
@albinpopote what's the status on this?

**Comment by thedoubl3j** [comment]
> @albinpopote what's the status on this?

waiting on @gundalow question. @angystardust thanks for testing on hardware. helps us out greatly

File: lib/ansible/modules/storage/netapp/na_ontap_net_subnet.py (hunk 5 above)
Comment (thedoubl3j) [reply lib/ansible/modules/storage/netapp/na_ontap_net_subnet.py]: ping @albinpopote @carchi8py thoughts?

**Comment by carchi8py** [comment]
I've got ahead and pinged @albinpopote on out Slack channel about making the copyright change.

Is it possible for one us to change it to @albinpopote if they don't respond?

**Comment by albinpopote** [comment]
Sorry, I was on holiday...
Fixed: I removed the wrong copyright.

**Review by thedoubl3j — APPROVED** [review]
comments addressed and copyright fixed, licensing correct. Tests look good and module was also tested on vendor per @angystardust. lgtm 👍

**Comment by thedoubl3j** [comment]
cc @thaumos

**Review by lonico — APPROVED** [review]
shipit

**Pull request merged**